CORS-3870: add validations for subnets field with AWS API

tthvo opened the pull request:
**Description**

This include validations for the vpc.subnets field with AWS API to conform to the specifications for subnet role assignment in the case of BYO subnets. The validation criteria can be found at jira ticket [0] and enhancement proposal [1].

For automatic role selection (i.e. no subnets have roles assigned), the installer rejects BYO VPC with untagged subnets (i.e. those without tag `kubernetes.io/cluster/<cluster-id>`) and suggests users to add a tag `kubernetes.io/cluster/unmanaged` to those subnets.

**Other changes**

Other related changes include:

- Refactored AWS API helper to be reusable without different query input (i.e. by vpcID, or subnetIDs) and migrated them to sdk v2 if possible.
- Refactored and improved test structures.


**References**

[0] https://issues.redhat.com/browse/CORS-3870
[1] https://github.com/openshift/enhancements/blob/master/enhancements/installer/aws-lb-subnet-selection.md

tthvo commented:
/jira refresh
/label platform/aws

tthvo commented:
/retest-required

tthvo commented:
/test aro-unit

tthvo commented:
/cc @patrickdillon

tthvo commented:
Job [ci/prow/e2e-aws-ovn-shared-vpc-edge-zones](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/9599/pull-ci-openshift-installer-main-e2e-aws-ovn-shared-vpc-edge-zones/1904801877363527680) is failing since there are untagged subnets in the VPC, which is new requirement introduced by these changes 😅 

/test terraform-okd-scos-images
/test e2e-aws-ovn-single-node

tthvo commented:
Job [ci/prow/e2e-aws-ovn-single-node](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/9599/pull-ci-openshift-installer-main-e2e-aws-ovn-single-node/1904944265784463360) failure seems unrelated as cluster install did succeed, but it is not byo-subnet install anyways. Retesting just in case 😅 

/test e2e-aws-ovn-single-node

mtulio commented:
> Job [ci/prow/e2e-aws-ovn-shared-vpc-edge-zones](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/9599/pull-ci-openshift-installer-main-e2e-aws-ovn-shared-vpc-edge-zones/1904801877363527680) is failing since there are untagged subnets in the VPC, which is new requirement introduced by these changes 😅
> 
> /test terraform-okd-scos-images /test e2e-aws-ovn-single-node

*Interesting*. Is this validation running before or after the installer tagging subnets[0]? I am wondering why non-edge subnets aren't in the list as apparently[1] (I can be wrong) any subnet have tags.

[0] FWIW: Recently we also fixed edge subnets not being tagged by cluster tag in BYO VPC - fixed in `main` branch.

[1]
vpc template: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_installer/9599/pull-ci-openshift-installer-main-e2e-aws-ovn-shared-vpc-edge-zones/1905778633968979968/artifacts/e2e-aws-ovn-shared-vpc-edge-zones/ipi-conf-aws-sharednetwork/artifacts/01_vpc.yaml
edge subnet template: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_installer/9599/pull-ci-openshift-installer-main-e2e-aws-ovn-shared-vpc-edge-zones/1905778633968979968/artifacts/e2e-aws-ovn-shared-vpc-edge-zones/ipi-conf-aws-sharednetwork/artifacts/01_vpc_99_subnet.yaml

tthvo commented:
>  Is this validation running before or after the installer tagging subnets[0]? 

I believe this validation runs before subnet tagging. Even then, the installer only checks for any other subnets that are not specified in the `install-config.yaml` and are untagged 😓 The required tag is `kubernetes.io/cluster/<cluster-id>` (i.e. we check for prefix `kubernetes.io/cluster/`).

From the template, looks like the tag is set as `cluster-name-public-zone-name` (ie. [Value: !Join ['-', [!Ref ClusterName, "public", !Ref ZoneName]]](https://raw.githubusercontent.com/openshift/installer/master/upi/aws/cloudformation/01_vpc_99_subnet.yaml)), which would not satisfy this requirement?

mtulio commented on pkg/asset/installconfig/aws/validation.go:
```diff
@@ -461,27 +532,187 @@ func validateMachineNetworksContainIP(fldPath *field.Path, networks []types.Mach
+
+// validateUntaggedSubnets ensures there are no additional untagged subnets in the BYO VPC.
+// An untagged subnet is a subnet without tag kubernetes.io/cluster/<cluster-id>.
+// Untagged subnets may be selected by the CCM, leading to various bugs, RFEs, and support cases.
```
>leading to various bugs, RFEs, and support cases. 
what about to mention the bug/RFE/EP section related to this topic?

tthvo replied:
Updated in the latest commit. I will squash those fixups when the PR is good to go :D

tthvo commented:
> yep, only tag:Name is set. IIUC those "extra" subnets are the public ones, created by template but not used in IC (as we don't want to create nodes on it in install-time) on CI job.

Ah opps, I misread the tag value for its key lol thanks 😓 awn, it is really missing the required tag key then.

>  CI job need to be updated to satisfy this requirement, or introduce a flag to the template to optionally create those public subnets.

Agreed with the suggestions! That means something like:

- CI step to add a tag `kubernetes.io/cluster/unmanaged` to those subnets after applying the template? 
- A condition in the template, lets call `create` (default to `true`), to allow skipping public subnets?

That would probably be captured in [CORS-3876](https://issues.redhat.com/browse/CORS-3876). I can update the ticket to reflect the discussions here :D

patrickdillon commented on pkg/asset/installconfig/aws/metadata.go:
```diff
@@ -250,7 +249,27 @@ func (m *Metadata) PublicSubnets(ctx context.Context) (Subnets, error) {
+	return m.subnets, nil
+}
+
+// VPCSubnets retrieves a group of all subnet metadata that is indexed by subnet ID
+// in the VPC of the provided subnets.
```
Having VPCSubnets in the metadata concerns me that it could be used accidentally when intending to use Subnets. The only purpose is for the validation right? Can you add a comment indicating that this may potentially include non-cluster subnets, and that Subnets should be used to select cluster subnets.

tthvo replied:
Sure, good idea! This field is indeed only for validations. Updated, thanks!

patrickdillon replied:
What did you decide as the resolution? The update is not obvious to me.

tthvo replied:
I added the following comment:

```
// vpcSubnets includes all subnets in the VPC.
// That is cluster subnets, which are provided in the install config and potentially other non-cluster subnets.
// Field subnets should be used to select only cluster subnets.
vpcSubnets    SubnetGroups
```

Maybe, I should also mention that it is only used for validation.

tthvo replied:
How does this https://github.com/openshift/installer/pull/9599/commits/50bbb98197edb5d58750c474047b2a4080bcefcf sound?

patrickdillon replied:
It sounds good, but I was thinking the comment would be on the func VPCSubnets as that is expected way for devs to interact with the metadata.

tthvo replied:
Lol now it makes sense why it was not obvious to you. My brain was not braining. Let me update it :sweat_smile:

tthvo replied:
Updated in https://github.com/openshift/installer/pull/9599/commits/3da57f1066b10d3bacd7cd674e03a532ec8033cf along with other comments :D

tthvo commented:
Updates: I extracted some previous miscellaneous fixes to https://github.com/openshift/installer/pull/9639 to reduce the scope of this PR :sweat_smile:

tthvo commented:
/retest

patrickdillon commented on pkg/asset/installconfig/aws/subnet.go:
```diff
@@ -42,92 +55,82 @@ type SubnetGroups struct {
+				return fmt.Errorf("%s has no VPC", *subnet.SubnetId)
+			}
+			if len(ptr.Deref(subnet.AvailabilityZone, "")) == 0 {
+				return fmt.Errorf("%s has not availability zone", *subnet.SubnetId)
```
nit: `not` -> `no`

mtulio commented:
> Agreed with the suggestions! That means something like:
> 
> * CI step to add a tag `kubernetes.io/cluster/unmanaged` to those subnets after applying the template?

I believe this one would makes more impact in the job [e2e-aws-ovn-shared-vpc-edge-zones](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/9599/pull-ci-openshift-installer-main-e2e-aws-ovn-shared-vpc-edge-zones/1909308083683725312),
and in those new validations, as it would check 1) edge subnets roles logic, 2) unused subnets (including edge) in the VPC should be tagged with the correct cluster tag `kubernetes.io/cluster/unmanaged`.

FWIW the change improving the conditional would be here: 
https://github.com/openshift/release/blob/8923214f0c11aaaaaf69f82280932dfba47133ec/ci-operator/step-registry/ipi/conf/aws/sharednetwork/ipi-conf-aws-sharednetwork-commands.sh#L218-L227

patrickdillon reviewed (commented):
Still working on the review. I have mostly worked through the subnet.go changes which are most tedious, so I should be able to finish up soon. I need to take a break so submitting the few comments I have so far.

This looks great so far.

mtulio commented:
I can't see the standard job for AWS BYO VPC (non-edge, used to be something like [`e2e-aws-ovn-shared-vpc`][job1] in the running list - I can see only in branches earlier than [release-4.16][job2].

Checking the job list:
/test ?

[job1]: https://prow.ci.openshift.org/job-history/gs/test-platform-results/pr-logs/directory/pull-ci-openshift-installer-main-e2e-aws-ovn-shared-vpc
[job2]: https://prow.ci.openshift.org/job-history/gs/test-platform-results/pr-logs/directory/pull-ci-openshift-installer-release-4.15-e2e-aws-ovn-shared-vpc

tthvo commented:
> I can't see the standard job for AWS BYO VPC (non-edge, used to be named by [`e2e-aws-ovn-shared-vpc`](https://prow.ci.openshift.org/job-history/gs/test-platform-results/pr-logs/directory/pull-ci-openshift-installer-main-e2e-aws-ovn-shared-vpc) ) in the running list - I can see only in branches earlier than [release-4.16](https://prow.ci.openshift.org/job-history/gs/test-platform-results/pr-logs/directory/pull-ci-openshift-installer-release-4.15-e2e-aws-ovn-shared-vpc).

It looks like [e2e-aws-ovn-shared-vpc-custom-security-groups](https://steps.ci.openshift.org/job?org=openshift&repo=installer&branch=main&test=e2e-aws-ovn-shared-vpc-custom-security-groups) is the closest? I do see [e2e-aws-ovn-shared-vpc](https://steps.ci.openshift.org/job?org=openshift&repo=installer&branch=main&test=e2e-aws-ovn-shared-vpc&variant=altinfra) but it is altinfra since 4.16.

Reference: https://steps.ci.openshift.org/search?job=e2e-aws-ovn-shared-vpc

tthvo commented:
Hmm, looks like there are some issues with test infra? E2E are failing with similar message:

```
level=fatal msg=failed to fetch Common Manifests: failed to fetch dependency of "Common Manifests": failed to generate asset "DNS Config": getting public zone for "origin-ci-int-aws.dev.rhcloud.com": no public route53 zone found matching name "origin-ci-int-aws.dev.rhcloud.com"
```

tthvo commented:
/retest

tthvo commented:
Updates: I added commit https://github.com/openshift/installer/pull/9599/commits/1d0463dfb7ebe31f264423a3c4ec916e0e63e764 and https://github.com/openshift/installer/pull/9599/commits/e193aa478009ea5f6de454c5ad6297ab01e72e1e (unit tests) to address the concern in [OCPBUGS-54832](https://issues.redhat.com/browse/OCPBUGS-54832).

tthvo commented:
/test e2e-aws-ovn-shared-vpc-edge-zones

patrickdillon commented:
> > I can't see the standard job for AWS BYO VPC (non-edge, used to be named by [`e2e-aws-ovn-shared-vpc`](https://prow.ci.openshift.org/job-history/gs/test-platform-results/pr-logs/directory/pull-ci-openshift-installer-main-e2e-aws-ovn-shared-vpc) ) in the running list - I can see only in branches earlier than [release-4.16](https://prow.ci.openshift.org/job-history/gs/test-platform-results/pr-logs/directory/pull-ci-openshift-installer-release-4.15-e2e-aws-ovn-shared-vpc).
> 
> It looks like [e2e-aws-ovn-shared-vpc-custom-security-groups](https://steps.ci.openshift.org/job?org=openshift&repo=installer&branch=main&test=e2e-aws-ovn-shared-vpc-custom-security-groups) is the closest? I do see [e2e-aws-ovn-shared-vpc](https://steps.ci.openshift.org/job?org=openshift&repo=installer&branch=main&test=e2e-aws-ovn-shared-vpc&variant=altinfra) but it is altinfra since 4.16.
> 
> Reference: https://steps.ci.openshift.org/search?job=e2e-aws-ovn-shared-vpc

yes the jobs were combined into one: https://github.com/openshift/release/pull/52445

patrickdillon commented on pkg/asset/installconfig/aws/validation.go:
```diff
@@ -461,24 +530,199 @@ func validateMachineNetworksContainIP(fldPath *field.Path, networks []types.Mach
+}
+
+// validateSubnetRoles ensures BYO subnets have valid roles assigned if roles are provided.
+func validateSubnetRoles(fldPath *field.Path, subnetsWithRole map[awstypes.SubnetRoleType][]subnetData, subnetDataGroups subnetDataGroups, publish types.PublishingStrategy) field.ErrorList {
```
If the cluster is `publish: External` the bootstrap node would have a public IP address, and therefore should go on a public subnet, I believe. 

I might have missed it, but if it's not there, do you think we should add it?

tthvo replied:
And also updated :D

tthvo commented on pkg/asset/installconfig/aws/validation.go:
```diff
@@ -461,24 +530,199 @@ func validateMachineNetworksContainIP(fldPath *field.Path, networks []types.Mach
+			continue
+		}
+
+		if ingressSubnet.Public != (publish == types.ExternalPublishingStrategy) {
```
Sure, I was seeing it is not supported in AWS so I skipped lol Let me add it. I got a sample [here](https://github.com/openshift/installer/pull/9633/files#diff-ab4fb360ac4a8344c506c58e1b799894584bcd0daf8f1216c118b11cc079d375) so I will copy it over.

tthvo replied:
Updated :D

patrickdillon replied:
> Sure, I was seeing it is not supported in AWS so I skipped lol 

Oh, I didn't realize it was currently only supported on Azure! Thanks for updating; this consistency will make adding support in other clouds a little bit easier :+1:

mtulio commented:
since this is change the subnet handling process, core of edge subnet classification, I would say the following job would be required to pass (it was good, but last run failed w/o clearly reasons):

/test e2e-aws-ovn-edge-zones

tthvo commented on pkg/asset/installconfig/aws/validation.go:
```diff
@@ -461,24 +530,199 @@ func validateMachineNetworksContainIP(fldPath *field.Path, networks []types.Mach
+
+	if len(untaggedSubnetIDs) > 0 {
+		errMsg := fmt.Sprintf("additional subnets %v without tag prefix %s are found in vpc %s of provided subnets. %s", untaggedSubnetIDs, TagNameKubernetesClusterPrefix, vpcSubnets.VPC,
+			fmt.Sprintf("Please add a tag %s to those subnets to exclude them from cluster installation or explicitly assign roles to provided subnets", TagNameKubernetesUnmanaged))
```
Updated :D

tthvo commented:
> since this is change the subnet handling process, core of edge subnet classification, I would say the following job would be required to pass (it was good, but last run failed w/o clearly reasons):

Sounds good to me! I added 2 commits to address Patrick's suggestions. PTAL again... I will squash those commits when we are happy :smile:

tthvo commented:
Ha, looks like the required tests passed :sunglasses: This PR will conflict with https://github.com/openshift/installer/pull/9544 in test file. I can rebase and squash commits if that PR got merged first.

mtulio commented:
> > since this is change the subnet handling process, core of edge subnet classification, I would say the following job would be required to pass (it was good, but last run failed w/o clearly reasons):
> 
> Sounds good to me! I added 2 commits to address Patrick's suggestions. PTAL again... I will squash those commits when we are happy 😄

> Ha, looks like the required tests passed 😎 This PR will conflict with https://github.com/openshift/installer/pull/9544 in test file. I can rebase and squash commits if that PR got merged first.

Amazing improvements here. 

/lgtm

I will hold based in your comments. Feel free to remove it when you are ready.

/hold

tthvo commented:
/hold cancel

I rebased and squashed the commits now :D The only change is in the test file to add a case for public-only cluster with no vpc subnets.

patrickdillon commented:
bumping Marco's

/lgtm

tthvo commented:
After #9633, there was some conflicts but the changes are additive. Rebased one more time :sweat_smile:
